transfer-sh.el

```diff
@@ -140,47 +149,167 @@ uploaded, otherwise the complete buffer is uploaded.
+                                transfer-sh-temp-file-location nil 0)
                                transfer-sh-temp-file-location)
-                           (if (buffer-file-name)
-                               buffer-file-name
-                             (progn
-                               (write-region (point-min) (point-max) transfer-sh-temp-file-location nil 0)
-                               transfer-sh-temp-file-location)))))
-    (if async
-        (transfer-sh-upload-file-async local-filename)
-      (transfer-sh-upload-file local-filename))))
+                           (or buffer-file-name
+                               (write-region
+                                (point-min)
+                                (point-max)
+                                transfer-sh-temp-file-location nil 0)
+                               transfer-sh-temp-file-location))))
+    (funcall (if async
+                 'transfer-sh-upload-file-async
+               'transfer-sh-upload-file)
+             local-filename)))
 
 ;;;###autoload
 (defun transfer-sh-upload-gpg (async)
-  "Uploads the active region/complete buffer to transfer.sh with gpg encryption.
+  "Allias for `tranfer-sh-encrypt-upload-region'."
+  (interactive "P")
+  (transfer-sh-encrypt-upload-region async))
+
+;;;###autoload
+(defun transfer-sh-encrypt-upload-region (async)
+  "Encrypt and upload the active region/complete buffer to transfer.sh.
 
-If a region is active, that region is encrypted using the
-user-given passcode and then uploaded, otherwise the complete
-buffer is encrypted and uploaded.  The argument given to gpg can
-be modifed using the transfer-sh-gpg-args variable."
+If a region is active, use that region, otherwise the complete
+buffer.
+
+Query user for the GPG key(s) to use for encryption. If no key is
+selected by user, then use symmetric encryption (and ask for a
+symmetric passphrase).
+
+The encrypted file is stored in `temporary-file-directory' and
+uploaded to transfer.sh using `transfer-sh-run-upload-agent'."
   (interactive "P")
-  (let* ((text (if (use-region-p) (buffer-substring-no-properties (region-beginning) (region-end)) (buffer-string)))
-    (cipher-text (substring
-     (shell-command-to-string
-      (concat
-       "echo " (shell-quote-argument text) "|"
-       "gpg --passphrase " (shell-quote-argument (read-passwd "Passcode: ")) " " transfer-sh-gpg-args))
-     0 -1))
-    (remote-filename (concat transfer-sh-remote-prefix (buffer-name) transfer-sh-remote-suffix)))
-    (save-excursion
-      (let* ((buf (find-file-noselect transfer-sh-temp-file-location)))
-	(set-buffer buf)
-	(erase-buffer)
-	(princ cipher-text buf)
-	(save-buffer)
-	(kill-buffer)))
-    (if async
-	(transfer-sh-upload-file-async transfer-sh-temp-file-location remote-filename)
-      (transfer-sh-upload-file transfer-sh-temp-file-location remote-filename))))
+  (or transfer-sh-gpg-keys-hash-table
+      (transfer-sh-refresh-gpg-keys))
```

Comment:
I'd use `unless` instead of `or` to better convey the meaning of what you want to do here.
`or` is more meaning full when setting variables or within conditions.